### ISSUE TITLE
no matching function for call to ‘atomic_store(std::atomic_int*, bool)’ in GCC 7

### ISSUE THREAD

**Issue by d-torrance** [Macaulay2/M2#3087]
My recent PR replacing libatomic_ops with std::atomic (#3082) broke the build when using GCC 7 (the default compiler in Ubuntu 18.04).  This is a pretty old compiler (GCC 8 was released in May 2018), and Ubuntu 18.04 reached the end of its standard support this past June, but it would be nice if Macaulay2 compiled on legacy systems.

The changes to the `d` and `e` directory compiled just fine -- it wasn't until we got to `system` that we ran into problems.  Maybe there's some easy fix?

```
g++ -std=gnu++14 -g -fdebug-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -Wformat -Werror=format-security -g3 -O2  -I. -I../d -I./../e -I/<<PKGBUILDDIR>>/M2/include -I/<<PKGBUILDDIR>>/M2/include -I/<<PKGBUILDDIR>>/M2/usr-host/include -I/<<PKGBUILDDIR>>/M2/include -I/<<PKGBUILDDIR>>/M2/include -I/<<PKGBUILDDIR>>/M2/usr-host/include -I/<<PKGBUILDDIR>>/M2/include -I/<<PKGBUILDDIR>>/M2/include -I/<<PKGBUILDDIR>>/M2/usr-host/include -Wdate-time -D_FORTIFY_SOURCE=2 -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular  -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include/python3.6m -I/usr/include -DBOOST_STACKTRACE_LINK -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular  -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include/python3.6m -I/usr/include -DBOOST_STACKTRACE_LINK -isystem /usr/include/libxml2   -DSING_NDEBUG -DOM_NDEBUG -DSING_NDEBUG -DOM_NDEBUG -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular -I/usr/include/x86_64-linux-gnu/singular -I/usr/include/singular  -I/usr/src/gtest  -Wdate-time -D_FORTIFY_SOURCE=2 -DNDEBUG  -I/usr/include/cdd -I/usr/include/eigen3 -I/usr/include/python3.6m -I/usr/include -DBOOST_STACKTRACE_LINK -I./../d -I./../c  -c -o supervisor.o supervisor.cpp
supervisor.cpp: In member function ‘void ThreadSupervisor::_i_cancelTask(ThreadTask*)’:
supervisor.cpp:339:67: error: no matching function for call to ‘atomic_store(std::atomic_int*, bool)’
       atomic_store(&task->m_CurrentThread->m_Interrupt->field,true);
                                                                   ^
In file included from /<<PKGBUILDDIR>>/M2/include/M2/atomic-field.h:5:0,
                 from supervisor.hpp:5,
                 from supervisor.cpp:1:
/usr/include/c++/7/atomic:1143:5: note: candidate: template<class _ITp> void std::atomic_store(volatile std::atomic<_ITp>*, _ITp)
     atomic_store(volatile atomic<_ITp>* __a, _ITp __i) noexcept
     ^~~~~~~~~~~~
/usr/include/c++/7/atomic:1143:5: note:   template argument deduction/substitution failed:
supervisor.cpp:339:67: note:   deduced conflicting types for parameter ‘_ITp’ (‘int’ and ‘bool’)
       atomic_store(&task->m_CurrentThread->m_Interrupt->field,true);
                                                                   ^
In file included from /<<PKGBUILDDIR>>/M2/include/M2/atomic-field.h:5:0,
                 from supervisor.hpp:5,
                 from supervisor.cpp:1:
/usr/include/c++/7/atomic:1138:5: note: candidate: template<class _ITp> void std::atomic_store(std::atomic<_ITp>*, _ITp)
     atomic_store(atomic<_ITp>* __a, _ITp __i) noexcept
     ^~~~~~~~~~~~
/usr/include/c++/7/atomic:1138:5: note:   template argument deduction/substitution failed:
supervisor.cpp:339:67: note:   deduced conflicting types for parameter ‘_ITp’ (‘int’ and ‘bool’)
       atomic_store(&task->m_CurrentThread->m_Interrupt->field,true);
                                                                   ^
In file included from /usr/include/c++/7/memory:82:0,
                 from /usr/include/c++/7/thread:39,
                 from supervisor.cpp:9:
/usr/include/c++/7/bits/shared_ptr_atomic.h:158:5: note: candidate: template<class _Tp, __gnu_cxx::_Lock_policy _Lp> void std::atomic_store(std::__shared_ptr<_Tp, _Lp>*, std::__shared_ptr<_Tp, _Lp>)
     atomic_store(__shared_ptr<_Tp, _Lp>* __p, __shared_ptr<_Tp, _Lp> __r)
     ^~~~~~~~~~~~
/usr/include/c++/7/bits/shared_ptr_atomic.h:158:5: note:   template argument deduction/substitution failed:
supervisor.cpp:339:67: note:   ‘std::atomic_int {aka std::atomic<int>}’ is not derived from ‘std::__shared_ptr<_Tp, _Lp>’
       atomic_store(&task->m_CurrentThread->m_Interrupt->field,true);
                                                                   ^
In file included from /usr/include/c++/7/memory:82:0,
                 from /usr/include/c++/7/thread:39,
                 from supervisor.cpp:9:
/usr/include/c++/7/bits/shared_ptr_atomic.h:143:5: note: candidate: template<class _Tp> void std::atomic_store(std::shared_ptr<_Tp>*, std::shared_ptr<_Tp>)
     atomic_store(shared_ptr<_Tp>* __p, shared_ptr<_Tp> __r)
     ^~~~~~~~~~~~
/usr/include/c++/7/bits/shared_ptr_atomic.h:143:5: note:   template argument deduction/substitution failed:
supervisor.cpp:339:67: note:   ‘std::atomic_int {aka std::atomic<int>}’ is not derived from ‘std::shared_ptr<_Tp>’
       atomic_store(&task->m_CurrentThread->m_Interrupt->field,true);
                                                                   ^
supervisor.cpp:340:67: error: no matching function for call to ‘atomic_store(std::atomic_int*, bool)’
       atomic_store(&task->m_CurrentThread->m_Exception->field,true);
                                                                   ^
In file included from /<<PKGBUILDDIR>>/M2/include/M2/atomic-field.h:5:0,
                 from supervisor.hpp:5,
                 from supervisor.cpp:1:
/usr/include/c++/7/atomic:1143:5: note: candidate: template<class _ITp> void std::atomic_store(volatile std::atomic<_ITp>*, _ITp)
     atomic_store(volatile atomic<_ITp>* __a, _ITp __i) noexcept
     ^~~~~~~~~~~~
/usr/include/c++/7/atomic:1143:5: note:   template argument deduction/substitution failed:
supervisor.cpp:340:67: note:   deduced conflicting types for parameter ‘_ITp’ (‘int’ and ‘bool’)
       atomic_store(&task->m_CurrentThread->m_Exception->field,true);
                                                                   ^
In file included from /<<PKGBUILDDIR>>/M2/include/M2/atomic-field.h:5:0,
                 from supervisor.hpp:5,
                 from supervisor.cpp:1:
/usr/include/c++/7/atomic:1138:5: note: candidate: template<class _ITp> void std::atomic_store(std::atomic<_ITp>*, _ITp)
     atomic_store(atomic<_ITp>* __a, _ITp __i) noexcept
     ^~~~~~~~~~~~
/usr/include/c++/7/atomic:1138:5: note:   template argument deduction/substitution failed:
supervisor.cpp:340:67: note:   deduced conflicting types for parameter ‘_ITp’ (‘int’ and ‘bool’)
       atomic_store(&task->m_CurrentThread->m_Exception->field,true);
                                                                   ^
In file included from /usr/include/c++/7/memory:82:0,
                 from /usr/include/c++/7/thread:39,
                 from supervisor.cpp:9:
/usr/include/c++/7/bits/shared_ptr_atomic.h:158:5: note: candidate: template<class _Tp, __gnu_cxx::_Lock_policy _Lp> void std::atomic_store(std::__shared_ptr<_Tp, _Lp>*, std::__shared_ptr<_Tp, _Lp>)
     atomic_store(__shared_ptr<_Tp, _Lp>* __p, __shared_ptr<_Tp, _Lp> __r)
     ^~~~~~~~~~~~
/usr/include/c++/7/bits/shared_ptr_atomic.h:158:5: note:   template argument deduction/substitution failed:
supervisor.cpp:340:67: note:   ‘std::atomic_int {aka std::atomic<int>}’ is not derived from ‘std::__shared_ptr<_Tp, _Lp>’
       atomic_store(&task->m_CurrentThread->m_Exception->field,true);
                                                                   ^
In file included from /usr/include/c++/7/memory:82:0,
                 from /usr/include/c++/7/thread:39,
                 from supervisor.cpp:9:
/usr/include/c++/7/bits/shared_ptr_atomic.h:143:5: note: candidate: template<class _Tp> void std::atomic_store(std::shared_ptr<_Tp>*, std::shared_ptr<_Tp>)
     atomic_store(shared_ptr<_Tp>* __p, shared_ptr<_Tp> __r)
     ^~~~~~~~~~~~
/usr/include/c++/7/bits/shared_ptr_atomic.h:143:5: note:   template argument deduction/substitution failed:
supervisor.cpp:340:67: note:   ‘std::atomic_int {aka std::atomic<int>}’ is not derived from ‘std::shared_ptr<_Tp>’
       atomic_store(&task->m_CurrentThread->m_Exception->field,true);
                                                                   ^
supervisor.cpp: In member function ‘void SupervisorThread::threadEntryPoint()’:
supervisor.cpp:435:45: error: no matching function for call to ‘atomic_store(std::atomic_int*, bool)’
       atomic_store(&m_Interrupt->field,false);
                                             ^
In file included from /<<PKGBUILDDIR>>/M2/include/M2/atomic-field.h:5:0,
                 from supervisor.hpp:5,
                 from supervisor.cpp:1:
/usr/include/c++/7/atomic:1143:5: note: candidate: template<class _ITp> void std::atomic_store(volatile std::atomic<_ITp>*, _ITp)
     atomic_store(volatile atomic<_ITp>* __a, _ITp __i) noexcept
     ^~~~~~~~~~~~
/usr/include/c++/7/atomic:1143:5: note:   template argument deduction/substitution failed:
supervisor.cpp:435:45: note:   deduced conflicting types for parameter ‘_ITp’ (‘int’ and ‘bool’)
       atomic_store(&m_Interrupt->field,false);
                                             ^
In file included from /<<PKGBUILDDIR>>/M2/include/M2/atomic-field.h:5:0,
                 from supervisor.hpp:5,
                 from supervisor.cpp:1:
/usr/include/c++/7/atomic:1138:5: note: candidate: template<class _ITp> void std::atomic_store(std::atomic<_ITp>*, _ITp)
     atomic_store(atomic<_ITp>* __a, _ITp __i) noexcept
     ^~~~~~~~~~~~
/usr/include/c++/7/atomic:1138:5: note:   template argument deduction/substitution failed:
supervisor.cpp:435:45: note:   deduced conflicting types for parameter ‘_ITp’ (‘int’ and ‘bool’)
       atomic_store(&m_Interrupt->field,false);
                                             ^
In file included from /usr/include/c++/7/memory:82:0,
                 from /usr/include/c++/7/thread:39,
                 from supervisor.cpp:9:
/usr/include/c++/7/bits/shared_ptr_atomic.h:158:5: note: candidate: template<class _Tp, __gnu_cxx::_Lock_policy _Lp> void std::atomic_store(std::__shared_ptr<_Tp, _Lp>*, std::__shared_ptr<_Tp, _Lp>)
     atomic_store(__shared_ptr<_Tp, _Lp>* __p, __shared_ptr<_Tp, _Lp> __r)
     ^~~~~~~~~~~~
/usr/include/c++/7/bits/shared_ptr_atomic.h:158:5: note:   template argument deduction/substitution failed:
supervisor.cpp:435:45: note:   ‘std::atomic_int {aka std::atomic<int>}’ is not derived from ‘std::__shared_ptr<_Tp, _Lp>’
       atomic_store(&m_Interrupt->field,false);
                                             ^
In file included from /usr/include/c++/7/memory:82:0,
                 from /usr/include/c++/7/thread:39,
                 from supervisor.cpp:9:
/usr/include/c++/7/bits/shared_ptr_atomic.h:143:5: note: candidate: template<class _Tp> void std::atomic_store(std::shared_ptr<_Tp>*, std::shared_ptr<_Tp>)
     atomic_store(shared_ptr<_Tp>* __p, shared_ptr<_Tp> __r)
     ^~~~~~~~~~~~
/usr/include/c++/7/bits/shared_ptr_atomic.h:143:5: note:   template argument deduction/substitution failed:
supervisor.cpp:435:45: note:   ‘std::atomic_int {aka std::atomic<int>}’ is not derived from ‘std::shared_ptr<_Tp>’
       atomic_store(&m_Interrupt->field,false);
                                             ^
```

Full log: https://launchpadlibrarian.net/711018721/buildlog_ubuntu-bionic-amd64.macaulay2_1.22.0.1+git202401242247-0ppa202401221544~ubuntu18.04.1_BUILDING.txt.gz

**Comment by moorewf** [comment]
It looks like the atomic fields are integers and one is passing in a bool value on line 339.  Perhaps just using 0 and 1 instead of true or false or explicitly casting to integers will fix the issue.

**Comment by d-torrance** [comment]
Yup, that did the trick -- thanks, Frank!

**Comment by d-torrance** [comment]
Closing -- this was fixed in #3088
